partiql-eval/src/eval/expr/operators.rs

```diff
@@ -0,0 +1,327 @@
+    Mod,
+    Exp,
+
+    // other ops
+    In,
+    Concat,
+}
+
+#[derive(Debug)]
+struct BoolShortCircuitArgChecker<const B: bool, OnMissing: ArgShortCircuit> {
+    marker: PhantomData<OnMissing>,
+}
+
+impl<const B: bool, OnMissing: ArgShortCircuit> ArgChecker
+    for BoolShortCircuitArgChecker<B, OnMissing>
+{
+    fn arg_check<'a>(
+        _typ: &PartiqlType,
+        arg: Cow<'a, Value>,
+    ) -> ArgCheckControlFlow<Value, Cow<'a, Value>> {
+        match arg.borrow() {
+            Boolean(b) if b == &B => ArgCheckControlFlow::ShortCircuit(Value::Boolean(*b)),
+            Missing => ArgCheckControlFlow::ShortCircuit(OnMissing::propagate()),
+            Null => ArgCheckControlFlow::Propagate(Null),
+            _ => ArgCheckControlFlow::Continue(arg),
+        }
+    }
+}
+
+impl BindEvalExpr for EvalOpBinary {
+    #[inline]
+    fn bind<const STRICT: bool>(
+        &self,
+        args: Vec<Box<dyn EvalExpr>>,
+    ) -> Result<Box<dyn EvalExpr>, BindError> {
+        type AndCheck = BoolShortCircuitArgChecker<false, PropagateNull<false>>;
+        type OrCheck = BoolShortCircuitArgChecker<true, PropagateNull<false>>;
+        type InCheck<const STRICT: bool> = DefaultArgChecker<STRICT, PropagateNull<false>>;
+        type Check<const STRICT: bool> = DefaultArgChecker<STRICT, PropagateMissing<true>>;
+        type EqCheck<const STRICT: bool> = DefaultArgChecker<STRICT, PropagateMissing<false>>;
+        type MathCheck<const STRICT: bool> = DefaultArgChecker<STRICT, PropagateMissing<true>>;
+
+        macro_rules! create {
+            ($check: ty, $types: expr, $f:expr) => {
+                BinaryValueExpr::create_checked::<{ STRICT }, $check, _>($types, args, $f)
+            };
+        }
+
+        macro_rules! logical {
+            ($check: ty, $f:expr) => {
+                create!($check, [TYPE_BOOL, TYPE_BOOL], $f)
+            };
+        }
+
+        macro_rules! equality {
+            ($f:expr) => {
+                create!(EqCheck<STRICT>, [TYPE_ANY, TYPE_ANY], $f)
+            };
+        }
+
+        macro_rules! math {
+            ($f:expr) => {{
+                let nums = PartiqlType::any_of([TYPE_INT, TYPE_REAL, TYPE_DOUBLE, TYPE_DECIMAL]);
+                create!(MathCheck<STRICT>, [nums.clone(), nums], $f)
+            }};
+        }
+
+        match self {
+            EvalOpBinary::And => logical!(AndCheck, |lhs, rhs| lhs.and(rhs)),
+            EvalOpBinary::Or => logical!(OrCheck, |lhs, rhs| lhs.or(rhs)),
+            EvalOpBinary::Eq => equality!(|lhs, rhs| NullableEq::eq(lhs, rhs)),
+            EvalOpBinary::Neq => equality!(|lhs, rhs| NullableEq::neq(lhs, rhs)),
+            EvalOpBinary::Gt => equality!(|lhs, rhs| NullableOrd::gt(lhs, rhs)),
+            EvalOpBinary::Gteq => equality!(|lhs, rhs| NullableOrd::gteq(lhs, rhs)),
+            EvalOpBinary::Lt => equality!(|lhs, rhs| NullableOrd::lt(lhs, rhs)),
+            EvalOpBinary::Lteq => equality!(|lhs, rhs| NullableOrd::lteq(lhs, rhs)),
```

Comment:
These comparison ops (`<`, `>`, `<=`, `>=`) have different behavior than the equality ops (`=`, `<>`) in strict mode. The equality ops should never error in strict mode and never return missing in permissive mode. While the comparison ops should return an error on non-comparable types and return missing in permissive mode.

Perhaps this could be a separate effort/issue to
- add conformance tests for this behavior
- fix this implementation

Reply:
https://github.com/partiql/partiql-lang-rust/issues/417

Reply:
Could also have #417 in the text?